freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncForEnvService.java

```diff
@@ -0,0 +1,155 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+import com.google.common.collect.Multimap;
+import com.sequenceiq.cloudbreak.common.json.Json;
+import com.sequenceiq.cloudbreak.logger.MDCUtils;
+import com.sequenceiq.freeipa.api.v1.freeipa.user.model.FailureDetails;
+import com.sequenceiq.freeipa.api.v1.freeipa.user.model.SuccessDetails;
+import com.sequenceiq.freeipa.api.v1.freeipa.user.model.SynchronizationStatus;
+import com.sequenceiq.freeipa.configuration.UsersyncConfig;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.entity.UserSyncStatus;
+import com.sequenceiq.freeipa.service.freeipa.user.model.SyncStatusDetail;
+import com.sequenceiq.freeipa.service.freeipa.user.model.UmsEventGenerationIds;
+import com.sequenceiq.freeipa.service.freeipa.user.model.UmsUsersState;
+import com.sequenceiq.freeipa.service.freeipa.user.model.UserSyncOptions;
+import com.sequenceiq.freeipa.service.freeipa.user.ums.UmsEventGenerationIdsProvider;
+import com.sequenceiq.freeipa.service.freeipa.user.ums.UmsUsersStateProviderDispatcher;
+import com.sequenceiq.freeipa.service.operation.OperationService;
+
+@Service
+public class UserSyncForEnvService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserSyncForEnvService.class);
+
+    @Inject
+    private UmsEventGenerationIdsProvider umsEventGenerationIdsProvider;
+
+    @Inject
+    private UserSyncForStackService userSyncForStackService;
+
+    @Inject
+    private UmsUsersStateProviderDispatcher umsUsersStateProviderDispatcher;
+
+    @Inject
+    @Qualifier(UsersyncConfig.USERSYNC_TASK_EXECUTOR)
+    private ExecutorService asyncTaskExecutor;
+
+    @Inject
+    private OperationService operationService;
+
+    @Inject
+    private UserSyncStatusService userSyncStatusService;
+
+    @Inject
+    private UmsVirtualGroupCreateService umsVirtualGroupCreateService;
+
+    public void synchronizeUsers(String operationId, String accountId, List<Stack> stacks, UserSyncRequestFilter userSyncFilter,
+            UserSyncOptions options) {
+        operationService.tryWithOperationCleanup(operationId, accountId, () -> {
+            Set<String> environmentCrns = stacks.stream().map(Stack::getEnvironmentCrn).collect(Collectors.toSet());
+            UserSyncLogEvent logUserSyncEvent = options.isFullSync() ? FULL_USER_SYNC : PARTIAL_USER_SYNC;
+            LOGGER.info("Starting {} for environments {} with operationId {} ...", logUserSyncEvent, environmentCrns, operationId);
+
+            if (options.isFullSync()) {
+                umsVirtualGroupCreateService.createVirtualGroups(accountId, environmentCrns);
```

Comment:
Why is this needed? Aren't the virtual groups created/deleted as part of the environment lifecycle? 
The virtual groups, a.k.a. workload administration groups, will be returned by the UMS in either the DefaultUmsUsersStateProvider or the BulkUmsUsersStateProvider.

Reply:
please see this commit: https://github.com/hortonworks/cloudbreak/commit/f9cab95e5346542b875e7e0572e6769e177a0e5b

Reply:
we need it to support newly introduced virtual groups for existing environments

Reply:
ok. this conversation isn't really related to this PR. let's move the discussion to the other JIRA